### ISSUE TITLE
pilot: Support configuring listeners with proxy protocol

### ISSUE THREAD

**Issue by tcolgate** [istio/istio#5384]
Ideally we'd be able to use proxy protocol on the http and https listeners of proxies acting as ingresses. This allows original client IP to be passed in from a suitable front end load balancer (e.g. via a suitably configured AWS ELB). Envoy supports it, and it appears to be configurable via the config file or the v2 API (heptio contour have already [added this,](https://github.com/heptio/contour/pull/156)

In additon the helm charts would need updating:

apiVersion: v1                                                                                                                                 
kind: Service
metadata:
  name: istio-ingress
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp     
    service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: '*'

**Comment by eduardobaitello** [comment]
@tcolgate Did you get it to work? How do you configure Envoy to use proxy-protocol?

**Comment by tcolgate** [comment]
Nope. Had too many other problems I refracting istio on our end, so didn't
even get started on this.

**Comment by eduardobaitello** [comment]
@rshriram @vadimeisenbergibm Can this issue be reopened? Or at least assign/label #8337 to someone or some milestone...

**Comment by gjcarneiro** [comment]
I can't believe some people say "Istio is the future" and it doesn't even support the basic PROXY protocol, which ingress-nginx has supported for years.  For me this is a blocker, I cannot possibly accept traffic into our prod cluster if I do not know exactly where it is coming from, for legal reasons.

**Comment by kuzaxak** [comment]
@vadimeisenbergibm  Are there any plans to fix this issue?

**Comment by mmerrill3** [comment]
just ran into this same issue, I was puzzled why my ELB load balancer was returning a 400 with istio 1.1.0.  Turns out it was the proxy-protocol setting on the ELB that I added, b/c I wanted to see the real client IP addresses to do rate limiting.  I guess that's not possible currently.

Removing this annotation from my ELB/service definition for the gateway resolved the issue, but this is a head scratcher.  The other thing to note is that it was very difficult to look at the ingress gateway logs or pilot logs to see why I was getting a 400 response.  I guess it's b/c the upstream pod doesn't support the protocol, which is the ingress gateway pod.  

service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: '*'

**Comment by mmerrill3** [comment]
I think something like this will work, although I'm still grasping all the moving parts.  

https://github.com/mmerrill3/istio/commit/0d476fea0f6b5386fb4e5f00a7c20bd932305f7c

**Comment by mmerrill3** [comment]
I'll have to build the filter into the template for the proxy.

**Comment by mmerrill3** [comment]
I tried the following, but still no luck with the 400 errors.  If anyone has time, can you point me in the right direction?

I see the new option for the listener_filter in the static configuration used by pilot inside istio-proxy container's directory /etc/istio/proxy/envoy.yaml

But, it looks like that doesn't make it over the the "json" file on the ingress gateway for some reason, I'm missing some conversion layer.  The name of the file on the ingress-gateway is envoy-rev0.json, not envoy.yaml

Looking for pointers.

https://github.com/mmerrill3/istio/commit/cb8bd392c190a5ead14a31b8ba5c2b15ab3d75e4

**Comment by mmerrill3** [comment]
I've stashed my work here, the plumbing is done to pass through the option to enable proxy protocol.  I've seen it enabled on the ingress gateway, and pilot.  I still receive an error when placing the ingress gateway behind a proxy-protocol enabled ELB, so to me, either the envoy version we are using does not support the proxy-protocol, or there is something internal to the envoy proxy that doesn't allow the proxy protocol to be used.

https://github.com/mmerrill3/istio/tree/collab-proxy-protocol

**Comment by mmerrill3** [comment]
I'll need to update pilot to send the listener with the listener filter set for the proxy-protocol for http and https.  That's the part I'm missing.  My working branch only fixes port 15090, not port 80's listener.  I didn't realize that listeners were discovered from pilot.

**Comment by mmerrill3** [comment]
I got this to work, only as a proof of concept, here:

https://github.com/mmerrill3/istio/commit/6fd46f540efae3d2e7fbcaf214ce81f4d408dbce

After I applied this patch, when I deployed the bookshop demo into my AWS accout with an ELB with proxy protocol enabled, I see 200 success responses.  No more 400 errors.

I updated pilot to inject the ListenerFilter for the proxy protocol when the lds service is asked for a gateway filter.

I don't think this is the istio way to do this.  I would think that the gateway crd could be enhanced to allow for a port to be marked as using the proxy protocol, as opposed to applying the proxy protocol to every inbound listener on a gateway.  But, at least it shows that it can be done.  

If we follow the crd route as well, then we don't have to deal with sending a command line parameter to pilot, and trying to push that configuration all the way down to the config generator.  

This was a great learning experience for me, now I think I really understand the guts of pilot and the discovery services, so it wasn't all in vain.
